packages/celltags-extension/src/celltag.tsx

```diff
@@ -0,0 +1,204 @@
+      return [];
+    }
+    const allTags = reduce(
+      cells,
+      (allTags: string[], cell) => {
+        const tags: string[] = (cell.getMetadata('tags') as string[]) ?? [];
+        return [...allTags, ...tags];
+      },
+      []
+    );
+    return [...new Set(allTags)].filter(tag => tag !== '');
+  }
+
+  private _emptyAddTag(target: HTMLInputElement) {
+    target.value = '';
+    target.style.width = '49px';
+  }
+
+  private _onAddTagKeyDown(
+    props: FieldProps,
+    event: React.KeyboardEvent<HTMLInputElement>
+  ) {
+    const input = event.target as HTMLInputElement;
+
+    if (event.ctrlKey) return;
+
+    if (event.key === 'Enter') {
+      this.addTag(props, input.value);
+    } else if (event.key === 'Escape') {
+      this._emptyAddTag(input);
+    }
+  }
+
+  private _onAddTagFocus(event: React.FocusEvent<HTMLInputElement>) {
+    if (!this._editing) {
+      (event.target as HTMLInputElement).blur();
+    }
+  }
+
+  private _onAddTagBlur(input: HTMLInputElement) {
+    if (this._editing) {
+      this._editing = false;
+      this._emptyAddTag(input);
+    }
+  }
+
+  private _onChange(event: React.ChangeEvent<HTMLInputElement>) {
+    if (!event.target.value) {
+      this._emptyAddTag(event.target);
+    } else {
+      const tmp = document.createElement('span');
+      tmp.className = 'add-tag';
+      tmp.textContent = event.target.value;
+      // set width to the pixel length of the text
+      document.body.appendChild(tmp);
+      event.target.style.width = tmp.getBoundingClientRect().width + 8 + 'px';
+      document.body.removeChild(tmp);
+    }
+  }
+
+  private _onAddTagClick(
+    props: FieldProps,
+    event: React.MouseEvent<HTMLElement>
+  ) {
+    const elem = (event.target as HTMLElement).closest('div');
+    const input = elem?.childNodes[0] as HTMLInputElement;
+    if (!this._editing) {
+      this._editing = true;
+      input.value = '';
+      input.focus();
+    } else if (event.target !== input) {
+      this.addTag(props, input.value);
+    }
+    event.preventDefault();
+  }
+
+  private _onTagClick(props: FieldProps, tag: string) {
+    const data = props.formData;
+    if (data.includes(tag)) {
+      data.splice(data.indexOf(tag), 1);
+    } else {
+      data.push(tag);
+    }
+
+    props.formContext.updateMetadata({ [props.name]: data }, true);
+  }
+
+  render(props: FieldProps): JSX.Element {
+    const allTags: string[] = this.pullTags();
+
+    return (
+      <div className="cell-tags">
+        <div className="jp-FormGroup-fieldLabel jp-FormGroup-contentItem">
+          Cell Tags
+        </div>
+        {allTags &&
+          allTags.map((tag: string, i: number) => (
+            <div
+              key={i}
+              className={`tag ${
+                props.formData.includes(tag) ? 'applied-tag' : 'unapplied-tag'
+              }`}
+              onClick={() => this._onTagClick(props, tag)}
+            >
+              <div className="tag-holder">
```

Comment:
Should all caps global consts instead of inline CSS class strings be used here?

Please see `shell.ts` for reference:
https://github.com/jupyterlab/jupyterlab/blob/82e1a74b61a41fbe5d65e8542053b749cdfe6ea0/packages/application/src/shell.ts#L39